ingest/expression_files/dense.py

```diff
@@ -0,0 +1,175 @@
+        if len(set(self.header)) != len(self.header):
+            self.error_logger.error(
+                "Duplicated header values are not allowed", extra=self.extra_log_params
+            )
+            return False
+        return True
+
+    def validate_gene_keyword(self):
+        """Validates that 'Gene' is the first value in header"""
+        # File is an R formatted file
+        if (self.header[-1] == '') and (self.header[0].upper() != 'GENE'):
+            pass
+        else:
+            # If not R formatted file then first cell must be 'GENE'
+            if self.header[0].upper() != 'GENE':
+                self.error_logger.error(
+                    f'First cell in header must be GENE not {self.header[0]}',
+                    extra=self.extra_log_params,
+                )
+                return False
+        return True
+
+    # @trace
+    # def preprocess(self):
+    #     """Determines if file is R-formatted. Creates dataframe (df)"""
+    #     csv_file, open_file_object = self.open_file(self.file_path)
+    #     dtypes = {'GENE': object}
+    #
+    #     # # Remove white spaces and quotes
+    #     header = [col_name.strip().strip('\"') for col_name in self.header]
+    #     # See if R formatted file
+    #     if (header[-1] == '') and (header[0].upper() != 'GENE'):
+    #         header.insert(0, 'GENE')
+    #         # Although the last column in the header is empty, python treats it
+    #         # as an empty string,[ ..., ""]
+    #         header = header[0:-1]
+    #     else:
+    #         header[0] = header[0].upper()
+    #     # Set dtype for expression values to floats
+    #     dtypes.update({cell_name: 'float' for cell_name in header[1:]})
+    #     self.df = self.open_file(
+    #         self.file_path,
+    #         open_as='dataframe',
+    #         names=header,
+    #         skiprows=1,
+    #         dtype=dtypes,
+    #         # chunksize=100000, Save for when we chunk data
+    #     )[0]
+
+    @trace
+    def transform(self):
+        """Transforms dense matrix into gene data model.
+        """
+        start_time = datetime.datetime.now()
+        self.error_logger.info('Starting run at ' +
+                               str(start_time), extra=self.extra_log_params)
+        print('Starting run at ' + str(start_time))
+        num_processed = 0
+        gene_models = []
+        data_arrays = []
+        for all_cell_model in self.set_data_array_cells(
+                self.header[1:], ObjectId()):
+            data_arrays.append(all_cell_model)
+        # Represents row as an ordered dictionary
+        for row in self.csv_file:
+            numeric_scores = list(
+                map(lambda x: round(float(x.strip().strip("\'")), 3), row[1:])
+            )
+            valid_expression_scores = []
+            cells = []
+
+            for idx, expression_score in enumerate(numeric_scores, 1):
+                if expression_score > 0:
```

Comment:
This should be `if expression_score != 0`.  In looking at the existing gene expression parsing code in the Ruby codebase, we only check if the value is _equal_ to zero, not greater.  While unusual, there are instances of negative values in dense matrices that have been normalized/processed in certain ways.